packages/core/contracts/financial-templates/contract-for-difference/ContractForDifference.sol

```diff
@@ -0,0 +1,183 @@
+import "../../common/interfaces/ExpandedIERC20.sol";
+import "../../common/interfaces/IERC20Standard.sol";
+
+import "../../oracle/interfaces/OracleInterface.sol";
+
+import "../../oracle/interfaces/FinderInterface.sol";
+import "../../oracle/interfaces/OptimisticOracleInterface.sol";
+import "../../oracle/interfaces/IdentifierWhitelistInterface.sol";
+
+import "../../oracle/implementation/Constants.sol";
+
+abstract contract ContractForDifference is Testable, Lockable {
+    using FixedPoint for FixedPoint.Unsigned;
+    using SafeMath for uint256;
+    using SafeERC20 for IERC20;
+    using SafeERC20 for ExpandedIERC20;
+    using Address for address;
+
+    enum ContractState { Open, ExpiredPriceRequested, ExpiredPriceReceived }
+    ContractState public contractState;
+
+    uint32 public expirationTimestamp;
+
+    uint256 public collateralPerPair;
+    uint256 public totalCollateral;
+
+    int256 public expiryPrice;
+
+    uint256 public expirationTokensForCollateral;
+
+    bytes32 public priceIdentifier;
+
+    IERC20 public collateralToken;
+    ExpandedIERC20 public longToken;
+    ExpandedIERC20 public shortToken;
+
+    FinderInterface public finder;
+
+    ContractForDifferenceFinancialProductLibrary public financialProductLibrary;
+
+    modifier preExpiration() {
+        require(getCurrentTime() < expirationTimestamp, "Only callable pre-expiry");
+        _;
+    }
+
+    modifier postExpiration() {
+        require(getCurrentTime() >= expirationTimestamp, "Only callable post-expiry");
+        _;
+    }
+
+    modifier onlyOpenState() {
+        require(contractState == ContractState.Open, "Contract state is not OPEN");
+        _;
+    }
+
+    constructor(
+        uint32 _expirationTimestamp,
+        uint256 _collateralPerPair,
+        address _longTokenAddress,
+        address _shortTokenAddress,
+        address _finderAddress,
+        bytes32 _priceIdentifier,
+        address _collateralAddress,
+        address _financialProductLibrary,
+        address _timerAddress
+    ) Testable(_timerAddress) {
+        finder = FinderInterface(_finderAddress);
+        require(_expirationTimestamp > getCurrentTime());
+        require(_getIdentifierWhitelist().isIdentifierSupported(_priceIdentifier));
+        require(_financialProductLibrary != address(0));
+
+        expirationTimestamp = _expirationTimestamp;
+        collateralPerPair = _collateralPerPair;
+        longToken = ExpandedIERC20(_longTokenAddress);
+        shortToken = ExpandedIERC20(_shortTokenAddress);
+        collateralToken = IERC20(_collateralAddress);
+        priceIdentifier = _priceIdentifier;
+
+        financialProductLibrary = ContractForDifferenceFinancialProductLibrary(_financialProductLibrary);
+    }
+
+    function mint(uint256 tokensToMint) public preExpiration() nonReentrant() returns (uint256 collateralUsed) {
+        collateralUsed = FixedPoint.fromUnscaledUint(tokensToMint).mul(collateralPerPair).rawValue;
```

Comment:
nit: should we be scaling tokensToMint? Seems like we should just cast it to a FixedPoint.Unsigned?